src/lib/http/http_server.pl

```diff
@@ -86,29 +85,14 @@
     format("Listening at port ~d\n", [Port]),
     accept_loop(Socket).
 
+% Register handlers
 register_handlers([]).
-register_handlers([get(Path, Handler)|Handlers]) :-
-    asserta(http_handler(get, Path, Handler)),
-    register_handlers(Handlers).
-register_handlers([post(Path, Handler)|Handlers]) :-
-    asserta(http_handler(post, Path, Handler)),
-    register_handlers(Handlers).
-register_handlers([put(Path, Handler)|Handlers]) :-
-    asserta(http_handler(put, Path, Handler)),
-    register_handlers(Handlers).
-register_handlers([patch(Path, Handler)|Handlers]) :-
-    asserta(http_handler(patch, Path, Handler)),
-    register_handlers(Handlers).
-register_handlers([head(Path, Handler)|Handlers]) :-
-    asserta(http_handler(head, Path, Handler)),
-    register_handlers(Handlers).
-register_handlers([delete(Path, Handler)|Handlers]) :-
-    asserta(http_handler(delete, Path, Handler)),
-    register_handlers(Handlers).
-register_handlers([options(Path, Handler)|Handlers]) :-
-    asserta(http_handler(options, Path, Handler)),
+register_handlers([Handler|Handlers]) :-
+    Handler =.. [Method, Path, Closure],
+    asserta(http_handler(Method, Path, Closure)),
```

Comment:
Please consider passing around the handlers as arguments instead of asserting clauses dynamically: Asserting clauses can have unintended effects for example when spawning multiple servers in parallel or consecutively, and complicates testing and debugging the code.

Reply:
I know but then I got into a problem I don't know how to solve. In the route matching I'm using variables, but if I use them as they're on the list, it only works for the first time, because later the value keeps unified in the list with the concrete request data.

Let me explain with a sample. If I have:
```
get(user/User, parameter(User))
```
 as a handler and I store it in a list, the first time (GET /usr/mthom) User will unify with "mthom" for example, and we will call parameter("mthom", Request, Response). The second time, User will still be "mthom" because they're in the same list, so doing a GET /user/triska will fail because "triska" \= "mthom". I want to keep the variable for every request, no matter how many times has been called before and by passing a list I don't know how to do it.

Maybe it's impossible and I need to think in another syntax for this type or parameters.

Reply:
You can use `copy_term/2` to copy the entire list of handlers, thus using fresh variables each time you are using the rules. For example:

<pre>
<b>?- copy_term(X/Y-Y, Copy).</b>
   Copy = _A/_B-_B.
</pre>